dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CrudControllerAdvice.java

```diff
@@ -489,14 +489,34 @@ public void servletExceptionHandler(ServletException ex) throws ServletException
     throw ex;
   }
 
-  @ExceptionHandler({
-    IllegalArgumentException.class,
-    SchemaPathException.class,
-    JsonPatchException.class
-  })
+  @ExceptionHandler({SchemaPathException.class, JsonPatchException.class})
+  @ResponseBody
+  public WebMessage handleBadRequest(Exception ex) {
+    return badRequest(ex.getMessage());
+  }
+
+  /**
+   * Handles {@link IllegalArgumentException} and prints the stack trace to standard error. {@link
+   * IllegalArgumentException} is used in DHIS 2 application code but also by various frameworks to
+   * indicate programming errors, so stack trace must be printed and not swallowed.
+   */
+  @ExceptionHandler(IllegalArgumentException.class)
   @ResponseBody
-  public WebMessage handleBadRequest(Exception exception) {
-    return badRequest(exception.getMessage());
+  public WebMessage illegalArgumentExceptionHandler(IllegalArgumentException ex) {
+    ex.printStackTrace();
+    return badRequest(ex.getMessage());
+  }
+
+  /**
+   * Handles {@link IllegalStateException} and prints the stack trace to standard error. {@link
+   * IllegalArgumentException} is used in DHIS 2 application code but also by various frameworks to
+   * indicate programming errors, so stack trace must be printed and not swallowed.
+   */
+  @ExceptionHandler(IllegalStateException.class)
+  @ResponseBody
+  public WebMessage illegalArgumentExceptionHandler(IllegalStateException ex) {
+    ex.printStackTrace();
```

Comment:
Same as above?

Reply:
Yes

Reply:
That's okay. Updated it now. All three places now use `log.error()`.